UnicornRender/video/source/Camera.cpp

```diff
@@ -12,120 +12,290 @@ namespace unicorn
+    CameraProjection(window, cameraProj),
+    m_fov(45.f),
+    m_fovLowerBound(44.f),
+    m_fovUpperBound(45.f),
+    m_znear(0.1f),
+    m_zfar(1000.f)
+{
+    UpdateProjection();
+}
+
+void PerspectiveCamera::UpdateProjection()
+{
+    m_cameraProjection = glm::perspective(m_fov,
+                                          m_aspect,
+                                          m_znear,
+                                          m_zfar);
+}
+
+void PerspectiveCamera::Scroll(float yoffset)
+{
+    SetFov(m_fov - yoffset);
+}
+
+void PerspectiveCamera::SetFov(float fov)
 {
     m_fov = std::max(std::min(fov, m_fovUpperBound), m_fovLowerBound);
-    m_dirtyProjection = true;
+    UpdateProjection();
+}
+
+UNICORN_EXPORT void PerspectiveCamera::SetZNear(float znear)
+{
+    m_znear = znear;
+    UpdateProjection();
+}
+
+UNICORN_EXPORT void PerspectiveCamera::SetZFar(float zfar)
+{
+    m_zfar = zfar; 
+    UpdateProjection();
+}
+
+UNICORN_EXPORT float PerspectiveCamera::GetFov() const
+{
+    return m_fov;
+}
+
+OrthographicCamera::OrthographicCamera(system::Window* window, glm::mat4& cameraProj) :
+    CameraProjection(window, cameraProj),
+    m_orthoScale(100.f),
+    m_scaledAspect(m_aspect / m_orthoScale),
+    m_right(static_cast<float>(window->GetSize().first) / 2),
+    m_top(static_cast<float>(m_pWindow->GetSize().second) / 2),
+    m_back(-1000.f),
+    m_front(1000.f)
+{
+    UpdateProjection();
 }
 
-void Camera::UpdateViewMatrix()
+void OrthographicCamera::UpdateProjection()
 {
-    m_matrices.m_view = glm::lookAt(m_position,
-                                    m_position + m_direction,
-                                    m_upVector);
+    m_right = static_cast<float>(m_pWindow->GetSize().first) / 2;
+    m_top = static_cast<float>(m_pWindow->GetSize().second) / 2;
+
+    float xScaled = m_right * m_scaledAspect;
+    float yScaled = m_top * m_scaledAspect;
+
+    m_cameraProjection = glm::ortho(-xScaled, xScaled,
+                                    -yScaled, yScaled,
+                                    m_back, m_front);
 }
 
-void Camera::UpdateProjectionMatrix()
+void OrthographicCamera::Scroll(float yoffset)
 {
-    m_matrices.m_projection = glm::perspective(m_fov,
-                                               m_aspect,
-                                               m_znear,
-                                               m_zfar);
+    SetScale(m_orthoScale + yoffset);
 }
+
+void OrthographicCamera::SetScale(float scale)
+{
+    m_orthoScale = std::max(scale, 0.1f);
+    m_scaledAspect = m_aspect / m_orthoScale;
+    UpdateProjection();
+}
+
+UNICORN_EXPORT float OrthographicCamera::GetScale() const
+{
+    return m_orthoScale;
+}
+
+Camera2DController::Camera2DController(glm::mat4& cameraView) : CameraController(cameraView)
+{
+}
+
+UNICORN_EXPORT void Camera2DController::MoveUp(float deltaTime)
+{
+    m_position += glm::vec3(0.0f, -deltaTime * m_speed, 0.0f);
```

Comment:
is it the same as in `CameraFPSController::MoveUp`? Would it make sense to move this to `CameraController`?

Reply:
Not every controller must change view matrix by moveup and so on. For example - orbital or 3rd person (bundled to some person)